tests/unit-tests/wayland/test_lifetime_tracker.cpp

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright © 2020 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <william.wold@canonical.com>
+ */
+
+#include "mir/wayland/wayland_base.h"
+
+#include <gtest/gtest.h>
+#include <gmock/gmock.h>
+
+namespace mw = mir::wayland;
+
+using namespace testing;
+
+class MockTracker : public virtual mw::LifetimeTracker
+{
+public:
+    // Expose protected method
+    void mark_destroyed() const
+    {
+        LifetimeTracker::mark_destroyed();
+    }
+};
+
+TEST(LifetimeTrackerTest, first_destroy_listener_id_is_not_zero)
+{
+    MockTracker tracker;
+    auto const id = tracker.add_destroy_listener([](){});
+    EXPECT_THAT(id.as_value(), Ne(0));
+}
+
+TEST(LifetimeTrackerTest, destroy_listener_is_called)
+{
+    int call_count = 0;
+    {
+        MockTracker tracker;
+        tracker.add_destroy_listener([&](){ call_count++; });
+    }
+    EXPECT_THAT(call_count, Eq(1));
```

Comment:
Maybe it would be more expressive to use a mock for a listener and have something like:

```c++
    EXPECT_CALL(mocklistener, callback()).Times(1);
```

Reply:
Done